### DISCUSSION TITLE
[Feature Request] Fill custom fields from machine learning

### DISCUSSION THREAD

**Discussion by niels-przybilla** [paperless-ngx/paperless-ngx#6932] [Feature Requests]
### Description

It would be great to have the feature that custom fields are filled by machine learning.
In my case that is for example Invoice numbers and the amount of the invoice.

**Comment by deeliightful** [comment]
I take it this is not a feature.

**Reply by shamoon** [reply]
Correct, hence the feature request

**Comment by SnoutBug** [comment]
With this you could also possibly combine custom fields and tags into "dynamic tags".

I usually use tags to either differenciate between companies, institutions and projects or use them to display the current processing state which can also currently be achieved with custom fields.

For every other case, where you would not need a value for a key, you could add a "constant" datatype to the custom fields (or dynamic tags) whose value is its key, which is then equal to the current implementation of tags.

This would open up many doors for further implementations and will also streamline the entire process of creating fitting tags for the properties you want to describe

**Comment by ctschach** [comment]
That would be exactly what I'm looking for. It doesn't matter if this is via machine learning or simple regex. Searching for "Amount" or "Rechnungsbetrag" and use the data behind to feed into a custom field would be awesome.

**Comment by andresledo** [comment]
Concretely this is a very specific need of mine, for example in the payrolls that the custom field is filled with the amount that is always in the same position. Docuware for example does this really well.

**Reply by SnoutBug** [reply]
I am surprised no FOSS DMS has this capability yet, while almost every other closed source company promotes this as their main feature. I think however that `paperless-ngx` will be the first to offer this ability since it seems the ground work has already been done.

I wonder what, except for time constraints, holds the implementation of this back?

**Reply by GHLasse** [reply]
The ai is probably only built for a fixed number of categories at the moment? It should be possible to expand this though.

**Reply by SnoutBug** [reply]
I looked at the source and it seems to implement this with the current solution, you would have to do at least two passes, one to assign the custom field and one to get the value of the field.
I'm guessing that each custom field would need its own trained model each and you would need one model to assign the custom field in the first place. The latter would probably just be a copy of the current implementation of tags?
For the former the current system would probably need to be extended as you said to allow dynamic training of models. Also the models in need to be trained would have to be selected by the documents list of custom fields first.
Maybe @shamoon can give hints on what the proper way to do this would be.

**Comment by murimaa** [comment]
It would be great if there were also other ways to pull custom field values for a document. For example meta tag in pdf or a barcode, or a custom script. In Finland all invoices include a barcode that has the amount and due date encoded in it, with some scripting it is possible to get those values automatically.

**Reply by vamagnus** [reply]
or even regex for me.

**Comment by shamoon** [comment]
Unfortunately there is not really anything that currently exists in paperless that is capable of this. The current ML stuff works on the principle that the potential values for a field (correspondent, tag etc) come from a list of discrete [existing] values.

What would be pretty straightforward to do would be to automatically assign/suggest the field itself (without the value). But, I wonder how much people would want that without the actual value part?

Either way, implementing some kind of mechanism to pull out data for a field would probably need to come from someone smarter than me.

**Reply by smkent** [reply]
How much of a stretch would it be to support ML learning/matching on Select-type custom fields, since those have predefined options similar to Corresponded/Document type/etc.?

**Reply by ju6ge** [reply]
I think on of the issues with custom field prediction apart from the complexity of some of the fields, is that custom fields do not exist for each document at the beginning, so in essence what you need is not just the ability to predict them but to assign the custom fields to the document. Those are solved problems, but it is not the default for the normal processing flow of a document.

Currently Paperless Processing runs at the very beginning doing: Correspondent prediction, Document Type Prediction, Date Prediction and Storepath prediction. Only after this has completed is it possible to assign custom fields to documents based on the results of the first processing stage via user defined workflows. So if paperless where to support this built in, that would require making the processing flow more interconnect which raises the complexity quite a bit and would require user configuration of the pipeline to a much greater degree. 

Anyway that is just my thoughts maybe `paperless` devs have a different take on this.

If you are interested in a somewhat self contained solution for custom-field filling, take a look at my project [paperless-llm-workflows](https://github.com/ju6ge/paperless-llm-workflows), it works completely offline and integrates nicely into user defined workflows ;)

**Comment by ctschach** [comment]
So how does the Date detection work? Does this use some ML?

Let’s say you have a bunch of invoices, wouldn’t it be possible to „detect“ the invoice amount and fill this into a custom field similarity the date detection?

From what I understood, you can detect a keyword (e.g.Invoice Number) but not any value behind it? So you could define the keyword together with the field name.

I know, this started to become geeky, but can’t we simply use some regular expressions from the plain text content to get field values? Something like:

`Invoice.*Number.*(.[0-9])`

**Reply by shamoon** [reply]
Date detection does in fact use a regex but it's quite different because dates follow a very expected format (with some variations of course). Yes, asking the user to write a regex could work but man, it sounds like a bad idea (hard to test, high technical barrier for the average user). I wouldn't be willing to introduce a feature that requires that (in the same way that regex is an option for matching but it's one of many).

And just for fun to illustrate, here's what could happen with that regex example you gave (yes, via chatGPT):

> Your provided regex, `Invoice.*Number.*(.[0-9])`, will look for text that contains the word "Invoice," followed by any number of any characters (`.*`), followed by "Number," and then capture a single character (any character due to `.`) that is immediately followed by a digit (`[0-9]`). 
> 
> While this can seem fine for extracting invoice numbers, unintended matches might happen in several scenarios due to the loose matching conditions. Here are some problematic examples:
> 
> ### 1. Overly broad matches:
> - **"Invoice blah blah Number @ 3 dogs"**
>   - Match: `@3` (captures `@` as `.` and `3` as `[0-9]`).
>   - Problem: Includes a non-numeric character preceding the digit.
> 
> ### 2. Matches unrelated patterns:
> - **"This is not related to Invoice Number # 9 but talks about something else"**
>   - Match: `#9` (captures `#` as `.` and `9` as `[0-9]`).
>   - Problem: Pulls from entirely irrelevant contexts or other sections of the text.
> 
> ### 3. Ignores context and meaning:
> - **"Incorrectly formatted invoice data: Invoice!!!Number?98 cats"**
>   - Match: `?9` (captures `?` and `9`).
>   - Problem: Misinterprets punctuation-laden strings.
> 
> ### 4. Matches partial or unintended text:
> - **"Invoice received. Numbered accounts 748 are all active."**
>   - Match: `d7` (captures `d` and `7`).
>   - Problem: Captures mid-sentence partial matches unrelated to actual invoice numbers.
> 
> ### 5. Matches single characters in unrelated spots:
> - **"Invoice missing details: Number- A7 of forms incomplete"**
>   - Match: `A7` (captures `A` as `.` and `7` as `[0-9]`).
>   - Problem: Fails to distinguish between a valid invoice number and unrelated strings.

**Reply by dodgex** [reply]
I would love to see a feature for this. I understand your point regarding the technical barrier for an average user. But this could be for example an advanced feature with a feature flag similar to the barcode stuff.

I am currently still in an evaluating phase but got paperless to decently detect invoices for e.g. my internet/phone bill, my domain registrations and my server hoster. all three of them have diffrent layouts to find for example the invoice value from.

e.g. "Summe: 30,12 €" or "Rechnungsbetrag (Brutto) 1,23 EUR". it would be awesome if i could use workflows to assign those values to my custom field. And as we can already do stuff like "Trigger: Document added; Has correspondent; has document type" and even a "content matichng pattern" (not sure if this supports regex or what kind of pattern?). With this we could make a workflow for each detected invoice type and if it matches e.g "Summe: [0-9]{1,4},[0-9]{2} €" to only process invoices with this patter matched. and then in Actions do the Assignment of custom fields (already possible) but again with a pattern like "Summe: ([0-9]{1,4},[0-9]{2}) €" (note the `(` and `)` here) to assing the value to the extracted/matched value. 

Of course this would require quiet some fine tuning on the users side, and maybe some  regex knowledge. but tbh with chat gpt nowadays this should be even possible for not so technical people. aaaand those who think they cant handle that feature, dont have to enable it in the first place, if it is a feature flag



---
ChatGPT on what regex one could use for a certain line in the invoice document:

> To extract the value `123,23` from the line `Rechnungsbetrag (Brutto) 123,23 EUR`, you can use the following regular expression:
> 
> ```regex
> Rechnungsbetrag \(Brutto\)\s+([\d.,]+)
> ```
> 
> ### Explanation:
> 1. **`Rechnungsbetrag \(Brutto\)`**:
>    - Matches the literal text `Rechnungsbetrag (Brutto)`. 
>    - The parentheses `(` and `)` are escaped with backslashes (`\`) because they are special characters in regex.
> 
> 2. **`\s+`**:
>    - Matches one or more whitespace characters (spaces, tabs, etc.) after `(Brutto)`.
> 
> 3. **`([\d.,]+)`**:
>    - Captures the desired numeric value into a group:
>      - `\d` matches any digit (0–9).
>      - `.` matches a literal period.
>      - `,` matches a literal comma.
>      - The `+` quantifier ensures it captures one or more of these characters in sequence.
> 
> ### Usage:
> - The numeric value `123,23` will be captured in the first capturing group (group 1). If your tool directly assigns captured groups to custom fields, you can use this group.
> 
> Let me know if you need further refinements!

**Comment by SpuriousGer** [comment]
I also think it would be great if custom fields would be "first class citizens" just like correspondent, date, etc.

I had a similar feature request for the suggestion feature that those fields have: https://github.com/paperless-ngx/paperless-ngx/discussions/7337

**Comment by cbapel** [comment]
One of the tools I worked with in the past did an excellent job handling custom field extraction. For the user, all that was needed was to activate the field to be extracted, right-click to select the PDF text preceding the value (placing an anchor, so to speak), and then left-click to select the value. I’m guessing that in the background, this was automatically generating regex patterns from the two captured text fields and other data, such as relative page position. These settings were then saved as a vendor template for that type of document, and machine learning was used to select the appropriate vendor template during processing. It was possible to create multiple templates for a vendor, such as contracts, invoices, delivery notes, etc.

I can see how generalizing field extraction, especially custom fields, with machine learning across a wide range of recipients and document types is challenging. However, once ML determines the recipient and document type, the regex can be applied much more specifically. 

The templates also allowed for triggering specific rules and workflows, such as converting the extracted regex strings to other formats. For example, I could extract the account number from an invoice and then convert the value to cost centers based on rules that "translate" one value to another under certain conditions.

Another tool I used provided the option to specify whether a custom field is always displayed or conditionally displayed (e.g., based on document type or a combination of recipient and document type).

For system configurations, it might make sense to create "system" workflows automatically that cannot be modified unless forcefully by admins—similar to default system rules in firewalls. Creating these workflows automatically would leverage existing functionality, maintain transparency for admins, and simplify user setups for features like custom field options. A similar approach could be applied to workflows tied to recipient templates for string translation, which could only be modified from within the templates or forcefully by admins.

**Reply by cbapel** [reply]
Documentation [Continia](https://docs.continia.com/en-us/continia-document-capture/business-functionality/documents-and-templates/working-with-templates/)

**Comment by NebulaDev80** [comment]
Hi all,

I found a 'hook solution' for regex extraction: 
https://gist.github.com/brunoamaral/258e73a488a75c8fb60a16b99119933b

 In my opinion the benefit of integration is quite huge and will hype the tool. Additionally I don't see the 'hard to test, high technical barrier for the average user' as the user does not have to use it.
In case the regex found more than one result, the user can define the position of the value in the result array.

Maybe the upper link is a good starting point for a regex solution. My suggestion is to start with a simple regex support and enhance the functionality with machine learning or other options that help to make the usage as easy as possible ( e.g. generate Regex from a Text field 'Context' and 'Search Value') .


best regards

and thanks a lot for the project up to now - good job!

**Reply by SpuriousGer** [reply]
Looks like a great solution.

One aspect I haven't found in the database, so don't know where it is stored, are the suggestions.
Regex is a great starting point, amending the list of suggestions using the API would be a solution that enhances it further. For most people though, Regex is a great solution.

I still have not really figured out how the fields like create date are handled differently or why it wouldn't be easy to just have more fields "like those".

**Reply by NebulaDev80** [reply]
As far as I found it, suggestions for the date are taken from metadata ( creation date ) and via regex document content ( here an array of dates is returned ) I think there is nothing in database itself.

**Comment by olegmeglin** [comment]
I can understand Shamoon's concerns—a solution for dynamic content is difficult to implement and would likely be error-prone. That said, I believe that at least for static values, such as those in a dropdown field or a true/false field, an automatic assignment mechanism is feasible.

I think it's important to distinguish between dynamic values (like the total amount or invoice number) and static values, which can already be represented using the current tag logic. Personally, I'm not a fan of having too many tags per document, as that tends to clutter the UI and make it less user-friendly.

Instead, I find the custom fields functionality extremely useful. For different document types, I use custom fields to capture information such as:

- **Direction:** A dropdown field (e.g., "Incoming" or "Outgoing")
- **Payment Status:** A dropdown field (e.g., "Open," "Paid," "Overdue")
- **Recurring:** A Boolean field (true/false)
- **Contract Status:** A dropdown field (e.g., "Active," "Terminated," "Expired")
- etc.

The reason I prefer to work with custom fields is that they provide a defined framework for the values to be used. This creates a consistent system, which is less error-prone. For these kinds of static values, it seems technically feasible to implement an automatic assignment—similar to how tags are currently handled.

**Reply by SpuriousGer** [reply]
I really like your suggestion. This could be a solution as well if it allows, like correspondent, etc., the setting of a field value even if it is not part of a dropdown.

**Reply by shamoon** [reply]
Yea the implementation is a bit complex but this is what I have thought about here, allow “auto” for fields with discrete values, regex (only) for others

**Reply by NebulaDev80** [reply]
I support that this is a useful functionality.

Anyway, I this this does not really hit the point of the original request. It is more a new Request having a kind of 'standard values' in case a custom field is created.
For example a payment status is something that ( in my opinion ) should not be set automatically by the system to 'Paied' because many other documents have this status already.

**Reply by olegmeglin** [reply]
> Anyway, I this this does not really hit the point of the original request.

You're right, but the whole discussion is already drifted from „using machine learning“ to static RegEx pattern. This also has nothing in common with classic ML, but the RegEx approach is also interesting for predefined values ​​from a custom field dropdown.

> It is more a new Request having a kind of 'standard values' in case a custom field is created.

Again, then, the hole "RegEx on custom Fields solution" is a new request, because predefined custom field values (like a dropdown or a boolean) also could use RegEx expressions as there logic to prefill values.

> For example a payment status is something that ( in my opinion ) should not be set automatically by the system to 'Paied'

First of all, you should never 100% trust the prefilled values (correspondent, issue date, tags, etc.). **You should always have a workflow to double-check if you rely on accurate data.** This is what the `Inbox`-Tag Workflow is for.

The payment status is a perfect example for a custom field that should be set by a system to an specific value. Here are the reasons why:

1. It is less time-consuming to correct an incorrectly declared value than to have to set it manually for each invoice.
1. If the whole automatically detection stuff does not work, I can still work with keywords or RegEx for a more accurate/customized detection.

**Reply by NebulaDev80** [reply]
Hi,

> You're right, but the whole discussion is already drifted from „using machine learning“ to static RegEx pattern. This also has nothing in common with classic ML, but the RegEx approach is also interesting for predefined values ​​from a custom field dropdown.

my suggestion for regex was to use the 'result' of machine learning / checked values and use these fields as a kind of template for choosing the right regex. The idea is that with the these fields the 'kind' of document is nearly clear and that for a matching with regex has a higher hit rate instead of using pure regex.

In this example only the document_type is used, but of course also more fields could be prechecked:
[Link to Discussion](https://github.com/paperless-ngx/paperless-ngx/discussions/5332#discussioncomment-12518398)
br

**Comment by ju6ge** [comment]
Hey guys,

I wanted this feature as well and have put significant time into coming up with a solution to this. I have just released the first version of my paperless extension.

How it works:
- Uses the API to find documents with unfilled custom fields
- Uses a local "large" language model to predict the values of the custom fields from all available data on the document … no cloud provider required, completely selfhostable. 
- Adds tags to the documents to indicate the processing status (still being processed, finished)

Expected workflow:
- Paperless processes documents
- Use Workflows to assign custom fields based on document_type and so on
- Run `paperless-field-extractor` to process the documents and fill the custom fields

You can check it out here https://github.com/ju6ge/paperless-field-extractor. Let me know what you think and what could be improved.

Cheers,
ju6ge

**Reply by Limerick-gh** [reply]
Could you share some information regarding required resources? Like minimum and recommended?

**Reply by ju6ge** [reply]
I have been able to run this on my laptop just on the integrated GPU with reasonable performance ;) … Per default I am using a quite small model around 4GB in size. It will also run on CPU but will take quite a bit longer per document, but it is an asyncronous process that runs in the background, so it might not matter depending on your use case ;)

If you want to test the inference speed for your setup there is a dry run mode where you can test how fast response generation is.

You can swap out the model for different ones as you desire. I have not have enough test cases to already have a good answer on how to generally tune your system. But I hope that if there is enough interest in my solution, then more answers will emerge :tada:

**Reply by olegmeglin** [reply]
At first glance, this looks promising. It would be great if there were a way to work with an existing ollama server via API :)

**Reply by ju6ge** [reply]
@olegmeglin not sure if that would be that beneficial. Ollama is just running llama.cpp in the background anyway, this project uses llama.cpp directly so any model that is compatible with ollama should just work here as well. Also specifically the way I am guiding the output of the model would be more challenging to implement, because I am using gbnf grammars which is as far as I am aware not supported for ollama. Also it would make the setup more complex. Here you just need the software and a model and everything will run, using ollama would introduce more complexity. The final reason I did not want to use ollama is that I am thinking about doing LoRa to finetune the model based on existing corpus of documents in paperless, adding learning so to speak. Because of this I decided against relying on Ollama.

**Reply by olegmeglin** [reply]
@ju6ge totally fair points, using llama.cpp directly gives you strict control and keeps the stack simple. 

My proposal targets the use case where many self-hosters already run an Ollama endpoint on a capable and often under-utilized GPU. Using the Ollama API would let Paperless reuse that capacity instead of running a second inference stack. 

Coexistence with Ollama on the same machine is technically possible, but both stacks keep their own models/VRAM and can compete for GPU resources.

I fully understand and agree with your arguments, they’re totally valid. I mainly wanted to highlight this use case.

**Reply by ju6ge** [reply]
I get where you are coming from ;) … I still think this is out of scope for now. 

I am also running ollama myself and actually my current deployment of my custom field solution is running on the same machine as well. And first even if I where to use the ollama server for this they still would compete for the same resources, Ollama is very much geared towards a single user setup. You can still run these two in parallel and in the end its a question of timing. My solution is not permanently using the GPU resources but only when there are documents to processes. So it depends on your volume of documents that are coming into paperless at any given moment. Is that really that much that you should be concerned about it colliding with your use of ollama? And if it is then using your ollama would not solve the problem, because then ollama would be bussy processing the documents instead of your requests.

But you raise the point that I should really test what happens when both run at the same time. Maybe when I have time to investigate this I will ;)
